---
title: "Running the linear model on the NCI-60 cell lines"
author: "Tara Eicher and Jalal K. Siddiqui, PhD"
date: "1/11/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

The NCI-60 cancer cell lines were developed as a drug screening tool focusing on a range of cancer types.  In this vignette, we compare drug scores of NCI-60 cancer cell lines (1, 2).

## Loading in IntLIM and Files

IntLIM, available from Github, can be installed as in the documentation.  Once IntLIM is installed, what is necessary is loading in the library.  

```{r}
rm(list = ls())
if(!require("devtools")){
  install.packages("devtools")
}
library("devtools")
if(!require("IntLIM")){
  install_github("ncats/IntLIM")
}
library(IntLIM)
library(rmarkdown)
```

For the NCI-60 cell lines, metabolomics and gene expression data were downloadable from the DTP website (https://wiki.nci.nih.gov/display/ncidtpdata/molecular+target+data).  The Metabolon data consisting of 353 metabolites and 58 cell lines with 177 technical replicates total was filtered for metabolites that had a median coefficient of variation of below 0.3.  The coefficient value was arbitrarily selected to filter out technical replicates having high variability.  The resulting metabolite abundance data set of 280 metabolites was subsequently log2 transformed. Probes from the Chiron Affymetrix U133 were mapped to genes using the Ensembl database hgu133.plus.db.  Probes mapping to more than one gene were removed.  In cases where more than one probe was matching to a given gene, only the probe with the highest mean expression was used for analysis.  This resulted in a total of 17,987 genes.   

This data has been formatted for IntLIM.  We load in the data as follows.  The nci60.csv meta file contains a list with the phenotypic data file, metabolite data file, gene expression data file, metabolite meta file, and gene expression meta file. The function __ShowStats()__ will give a summary of the NCI-60 data for all samples containing both gene expression and metabolite abundance. We find that we have gene expression data involving 17,987 genes and metabolite abundance data involving 280 metabolites with 57 cell lines.

```{r}
csvfile <- file.path(getwd(), "/NCI60_data/nci60.csv")
inputData <- IntLIM::ReadData(inputFile = csvfile,analyteType1id='id',analyteType2id='id',
                              class.feat = list(drugscore = "numeric"))
IntLIM::ShowStats(inputData)
```

## Filtering Gene Expression and Metabolite Data

We remove genes with mean belows below the 10th percentile.  Furthermore, we remove analytes with more than 80% missing values (where an NA value indicates missingness).  This results in gene expression data involving 16,188 genes and metabolite abundance data involving 280 metabolites in 57 cell lines.  

```{r}
inputDatafilt <- IntLIM::FilterData(inputData,analyteType1perc=0.10, analyteMiss = 0.80)
IntLIM::ShowStats(inputDatafilt)
```

We can obtain boxplot distributions of the data as follows.

```{r}
IntLIM::PlotDistributions(inputDatafilt)
```

## Principal Component Analysis

The principal component analysis is performed on filtered metabolite and gene expression data to obtain visual representations showing how different sub-sections of the data could be grouped into different clusters.  Common samples are shown.  Darker blue samples indicate higher drug scores.
```{r}
PlotPCA(inputDatafilt, stype = "drugscore")
```

## Running Linear Model

The linear model for integrating multi-omics and data is:  a_i = β0 + β1 a_j + β2 p + β3 (a_j:p) + Σ(β(3+c) C) + ε where ‘a_i’ and ‘a_j’ are (optionally log-transformed) analyte abundances, ‘p’ is phenotype (cancer type, patient diagnosis, treatment group, etc), ‘(a_j:p)’ is the association between analyte level and phenotype, C is a set of clinical covariates (e.g. age, sex, etc), c is a covariate index, and ‘ε’ is the error term that is normally distributed.  A statistically significant p-value of the ‘(a_j:p)’ association term indicates that the slope relating analyte levels differs by phenotype. We run a linear model that included 16,188 genes and 280 metabolites (total of 4,532,640 possible associations and hence models).  The model is run as below by calling __RunIntLim()__.  __DistPvalues()__ allows us to obtain a distribution of p-values for the (a_j:p) term, and __DistRSquared()__ allows us to obtain a distribution of R^2 values for the models. __InteractionCoefficientGraph()__ allows us to obtain an ordered plot of the (a_j:p) term coefficients.  We also show a volcano plot showing the relationship between interaction coefficients and unadjusted p-values using __pvalCoefVolcano()__.

Covariates (e.g. age and batch) may also be included in the models as a list using the
__covar__ parameter. Please be advised that not all covariates in the experiment need be included.
In particular, colliders (variables influenced by one or more other variables) should not be
included, and potential confounding variables should be included. The choice of covariates
should be determined by the experiment.

Note that the user must select either analyte type 1 or analyte type 2 for the 
independent and outcome variables. This choice should be biologically motivated.
For instance, if the user wants to investigate metabolite abundances that are 
influenced by gene expression level and metabolite abundance is analyte type 1 and
gene expression is analyte type 2, then the outcome would be analyte type 1 and
the independent variable would be analyte type 2.

```{r}
myres <- IntLIM::RunIntLim(inputDatafilt, stype="drugscore", continuous = TRUE,
                           independent.var.type = 1, outcome = 2, save.covar.pvals = TRUE)
IntLIM::DistPvalues(myres, adjusted = FALSE)
IntLIM::DistRSquared(myres)
IntLIM::InteractionCoefficientGraph(inputResults = myres, 
                                    interactionCoeffPercentile = 0.9)
IntLIM::pvalCoefVolcano(inputResults = myres, inputData = inputDatafilt,
                        pvalcutoff = 0.05)
```

The next step is to process the results of this model by filtering the results of the linear model by FDR-adjusted p-value cutoff (0.10 selected here) for the (a_j:p) association coefficient. We further may only interested in results where the interaction coefficient is in the top Xth percentile (e.g. 50th shown here) and where the model fit as measured by R^2 value is above a given cutoff (e.g. 0.2). This is done with the __ProcessResults()__ function.

It is also possible to filter for significant main effects (independent analyte level
or phenotype). Users can then determine which significant pairs also have significant main effects.
```{r}
myres.sig <- IntLIM::ProcessResults(myres,  inputDatafilt, pvalcutoff = 0.10, rsquaredCutoff = 0,
                                    coeffPercentile = 0.5)
myres.sig.stype <- IntLIM::ProcessResults(myres,  inputDatafilt, pvalcutoff = 0.10, rsquaredCutoff = 0,
                                    coeffPercentile = 0.5, coefficient = "stype")
myres.sig.a <- IntLIM::ProcessResults(myres,  inputDatafilt, pvalcutoff = 0.10, rsquaredCutoff = 0,
                                    coeffPercentile = 0.5, coefficient = "analyte")
OutputResults(inputResults = myres.sig, filename = "NCI60pairs.csv")
```

From this model we find 2,517 gene-metabolite correlations that meet our criteria. From the __HistogramPairs()__ histogram plots below, we can see that most genes are associated with a single metabolite and that most metabolites are associated with a single gene. However, we see that some
genes are associated with up to 12 metabolites and that some metabolites are associated with up to
359 genes.

We can see that a few of the metabolites associated with high numbers of genes by drug score are 2'-deoxyuridine 5'-triphosphate, inositol 1-phosphate, beta-nicotinamide adenine dinucleotide, and cholesterol.
```{r}
# Plot histogram.
IntLIM::HistogramPairs(myres.sig, type = "outcome")
IntLIM::HistogramPairs(myres.sig, type = "independent")

# Tabulate gene and metabolite counts.
table.gene.count <- as.data.frame(table(as.character(myres.sig$Analyte1)))
table.metab.count <- as.data.frame(table(as.character(myres.sig$Analyte2)))
print(table.gene.count[order(-table.gene.count$Freq)[1:10],])
print(table.metab.count[order(-table.metab.count$Freq)[1:10],])

# Print gene associations with metabolites of interest.
str(myres.sig[which(myres.sig$Analyte2 == "X2..deoxyuridine.5..triphosphate"),])
str(myres.sig[which(myres.sig$Analyte2 == "inositol.1.phosphate"),])
str(myres.sig[which(myres.sig$Analyte2 == "beta.nicotinamide.adenine.dinucleotide"),])
str(myres.sig[which(myres.sig$Analyte2 == "cholesterol"),])
```

We can show some example plots of some of these pairs.  The first example is the HHAT vs. cholesterol.

The __PlotPairResiduals()__ function plots the standardized residuals for a given
pair, allowing the user to determine whether model assumptions are fulfilled. In
the example below, there is a slight linear trend in the residuals, indicating that
the relationship between HHAT and cholesterol may better be approximated by a more
complex model.

```{r}
IntLIM::PlotPair(inputDatafilt, myres, 
                 independentAnalyteOfInterest = "HHAT", 
                 outcomeAnalyteOfInterest = "cholesterol",
                 independentVariable = 1,
                 outcome = 2)
IntLIM::PlotPairResiduals(inputDatafilt, myres, 
                 independentAnalyteOfInterest = "HHAT", 
                 outcomeAnalyteOfInterest = "cholesterol",
                 independentVariable = 1,
                 outcome = 2)
```

Another example is CDKN1A vs. cholesterol. In
the example below, there is a linear trend in the residuals, indicating that
the relationship between CDKN1A and cholesterol may better be approximated by a more
complex model.

```{r}
IntLIM::PlotPair(inputDatafilt, myres, 
                 independentAnalyteOfInterest = "CDKN1A", 
                 outcomeAnalyteOfInterest = "cholesterol",
                 independentVariable = 1,
                 outcome = 2)
IntLIM::PlotPairResiduals(inputDatafilt, myres, 
                 independentAnalyteOfInterest = "CDKN1A", 
                 outcomeAnalyteOfInterest = "cholesterol",
                 independentVariable = 1,
                 outcome = 2)
```

## Cross-Validation
We then run 5-fold cross-validation to determine the overlap of significant pairs across folds. To replicate leave-one-out cross-validation as shown in the paper, change __folds__ from 5 to 57.
```{r results = FALSE}
folds <- 5
crossValResults <- RunCrossValidation(inputData = inputData, analyteType1perc = 0.10, 
                   analyteMiss = 0.80, stype="drugscore", outcome = c(2), 
                   independent.var.type = c(1), pvalcutoff = 0.10, 
                   rsquaredCutoff = 0.2,
                   folds = folds, continuous = TRUE,
                   interactionCoeffPercentile = 0.5)
IntLIM::PlotFoldOverlapUpSet(crossValResults$processed)

# Find number of pairs overlapping in at least half of the folds.
result <- lapply(1:folds, function(j){return(rownames(crossValResults$processed[[j]]))})
result_flat <- unlist(result)
fold_count <- table(result_flat)
length(which(fold_count > folds / 2))

# Find number of pairs from original data that overlap the cross-validation data.
resultPairs <- paste(myres.sig$Analyte1, myres.sig$Analyte2, sep = "__")
foldCountResults<- fold_count[which(names(fold_count) %in% resultPairs)]
str(foldCountResults)
length(which(foldCountResults > folds / 2))
```
## Permutation
Finally, we run 5 permutations of the data. To replicate the results of the paper, change
the num.permutations from 5 to 100. For large datasets and high numbers of permutations,
you may wish to run your code on a supercomputer. For very large datasets, you may wish to 
create several batch jobs on a supercomputer, each of which calls __PermuteIntLIM()__ with 
num.permutations = 1, and then concatenate the results after all jobs have completed. If you do 
this, it is recommended to change the seed for each call.

__PermuteIntLIM()__ returns R^2 values averaged over all pairs, the number of significant
pairs per permutation, and the significant pairs per permutation. These can be further
refined using __PermutationCountSummary()__, which computes the number of significant pairs, independent
variable analytes, and outcome analytes per permutation. Furthermore, a violin plot can
be generated by setting plot = TRUE. In addition, __PermutationPairSummary()__ computes
the number of permutations in which each pair was found to be significant. This can
also be visualized using a horizontal bar plot. For the NCI-60 data set, there are 4 pairs that
were significant in 4 out of 5 permutations. In addition, the number of metabolites
and pairs that were significant in the original data was higher than the number in
the permuted data.
```{r results = FALSE}
# Permute and print summary.
perm.res <- IntLIM::PermuteIntLIM(data = inputDatafilt, stype = "drugscore", outcome = 2,
                     independent.var.type = 1, pvalcutoff = 0.10, interactionCoeffPercentile = 0.5,
                  rsquaredCutoff = 0.2,
                  num.permutations = 5, continuous = TRUE)
summaryCount <- IntLIM::PermutationCountSummary(permResults = perm.res, inputResults = myres.sig,
                              plot = TRUE)
print(summaryCount)

# Compute the number of times each pair significant in the original data was
# also significant in the permuted data.
summaryPairs <- IntLIM::PermutationPairSummary(permResults = perm.res, inputResults = myres.sig,
                              plot = TRUE)
str(summaryPairs)
```

The list of unique genes and metabolites in these pairs can be used to conduct a pathway enrichment analysis. Using RaMP, we query the set of pairs for shared pathways and shared reactions.
```{r results = FALSE}
if(!require("RaMP")){
 install_github("ncats/RAMP-DB")
}
library(RaMP)

# Load files.
fdata <- inputDatafilt@analyteType2MetaData

# Match metabolite ID's to source.
metabid <- fdata[myres.sig$Analyte2, "databaseId"]

# Match gene ID's to source.
geneid <- paste0("gene_symbol:", myres.sig$Analyte1)

# Connect to RaMP. Replace with your MySQL password, or "" if not using a password.
password <- unlist(unname(read.csv(file.path("~/mysql_pw.txt"), header = FALSE)))
pkg.globals <- setConnectionToRaMP(dbname = "ramp", username = "root", conpass = password,
  host = "localhost")

# Find which pairs share pathways.
sharesPathway <- lapply(1:nrow(myres.sig), function(i){
  shares <- FALSE
  pwayResult <- getPathwayFromAnalyte(c(metabid[[i]], geneid[[i]]))
  pwayResultMetab <- pwayResult$pathwayId[which(pwayResult$inputId == metabid[[i]])]
  pwayResultGene <- pwayResult$pathwayId[which(pwayResult$inputId == geneid[[i]])]
  if(length(intersect(pwayResultMetab, pwayResultGene)) > 0){
    shares <- TRUE
  }
  return(shares)
})

# Find which pairs share reactions.
sharesRxn <- lapply(1:nrow(myres.sig), function(i){
  shares <- FALSE
  tryCatch({
    rxnResult <- rampFastCata(metabid[[i]])$rxn_partner_ids
    rxnResultAll <- unlist(lapply(rxnResult, function(res){
      return(strsplit(res, "; ")[[1]])
    }))
    if(geneid[[i]] %in% rxnResultAll){
      shares <- TRUE
    }
  }, error = function(cond){})
  return(shares)
})

```

```{r}
# Further analyses reported in paper for comparison with original data
print(summaryPairs[which(summaryPairs$Pair == "HHAT__V__cholesterol"),])
print(summaryPairs[which(summaryPairs$Pair == "CDKN1A__V__cholesterol"),])

# Compare to pairs that share pathways.
pathwayPairs <- paste(myres.sig[which(sharesPathway == TRUE), "Analyte1"], 
                        myres.sig[which(sharesPathway == TRUE), "Analyte2"], sep = "__V__")
length(which(summaryPairs[which(summaryPairs$Pair %in% pathwayPairs), "Perm.Count"] == 0))

# Compare to pairs that share reactions.
rxnPairs <- paste(myres.sig[which(sharesRxn == TRUE), "Analyte1"], 
                        myres.sig[which(sharesRxn == TRUE), "Analyte2"], sep = "__V__")
length(which(summaryPairs[which(summaryPairs$Pair %in% rxnPairs), "Perm.Count"] == 0))
```

This chunk evaluates a subset of pairs using glm, illustrating the difference in running time
between glm and IntLIM. This evaluation was conducted strictly for publication purposes and isn't
something the typical user will need to run. Nevertheless, the code is included here for
those who wish to replicate the results of the paper.
```{r eval = FALSE}
library(stats)
library(rsq)

# Subset data.
inputDatafilt@analyteType1 <- inputDatafilt@analyteType1[1:10,]

# Run IntLIM.
myres <- IntLIM::RunIntLim(inputDatafilt, stype="drugscore", continuous = TRUE,
                           independent.var.type = 1, outcome = 2)

# Run GLM.
starttime <- Sys.time()
subdat <- NULL
suppressMessages({pvals = lapply(rownames(inputDatafilt@analyteType2), function(m){
  pvals_single = lapply(rownames(inputDatafilt@analyteType1), function(g){
    subdat = data.frame(drugscore=inputDatafilt@sampleMetaData,
                        m=inputDatafilt@analyteType2[m,],
                        g=inputDatafilt@analyteType1[g,])
    attach(subdat)
    form_glm = "m ~ g + drugscore + g:drugscore"
    myfit = glm(form_glm, data = subdat)

    pvals = coef(summary(myfit))[,4]
    result = data.frame(rsq(myfit), pvals["g"], pvals["g:drugscore"],
                        pvals["drugscore"])
    colnames(result) = c("Rsquared", "Pval_gene", "Pval_interaction", "Pval_drugscore")
    rownames(result) = paste(m, g, sep = "_")
    return(result)
  })
  result_metab = do.call("rbind", pvals_single)
})})
endtime <- Sys.time()
print(endtime - starttime)
```

### References

1.  Su, G., Burant, C.F., Beecher, C.W., Athey, B.D. and Meng, F. (2011) Integrated metabolome and transcriptome analysis of the NCI60 dataset. BMC bioinformatics, 12, S36.

2.  Reinhold, W. C., Sunshine, M., Liu, H., Varma, S., Kohn, K. W., Morris, J., Doroshow, J., &#38; Pommier, Y. (2012). CellMiner: a web-based suite of genomic and pharmacologic tools to explore transcript and drug patterns in the NCI-60 cell line set. Cancer Research, 72(14), 3499. https://doi.org/10.1158/0008-5472.CAN-12-1370.