# Description
This directory includes two datasets formatted for use with IntLIM and sample vigettes for each respective dataset. Both datasets are properly formatted and ready to run.

# NCI-60
The NCI-60 data is located in `NCI60_data`, and the vignette is in `IntLIM2.0Vignette-NCI60.Rmd`. For the NCI-60 cell lines, metabolomics and gene expression data were downloadable from the DTP website (https://wiki.nci.nih.gov/display/ncidtpdata/molecular+target+data).  The Metabolon data consisting of 353 metabolites and 58 cell lines with 177 technical replicates total [1] was filtered for metabolites that had a median coefficient of variation of below 0.3.  The coefficient value was arbitrarily selected to filter out technical replicates having high variability.  The resulting metabolite abundance data set of 280 metabolites was subsequently log2 transformed. Probes from the Chiron Affymetrix U133 were mapped to genes using the Ensembl database hgu133.plus.db.  Probes mapping to more than one gene were removed.  In cases where more than one probe was matching to a given gene, only the probe with the highest mean expression was used for analysis.  This resulted in a total of 17,987 genes.   
The outcome is a continuous phenotype called the drug score. The drug score of each cell line was calculated in [2] using the following steps: (1) Evaluate the IC50 score of each drug for each cell line after 48 hours, (2) z-scale all IC50 scores, and (3) Compute the average z-score for each sample as a representative metric of drug responsiveness. In the vignette, the goal is to find gene-metabolite pairs that are associated with one another with respect to drug score.

# BRCA
A previous study had conducted both gene expression and metabolomics profiling of tissue samples from breast cancer patients (3).  This vigenette will highlight the analysis we conduct on the breast cancer data.  For breast cancer study, both gene expression (available on Gene Expression Omnibus Accession Number:  GSE37751) and metabolomics(1)  (http://www.jci.org/articles/view/71180/sd/2) data are available online.   Much of this data had been processed as previously described (1).  Probes from the Affymetrix data not mapping to a gene symbol were removed.  Additionally, as with NCI-60 data, only the probe corresponding to the highest mean expression was used for analysis when multiple probes corresponded to a single gene. This resulted in a total of 20,254 genes for 108 patient samples.  The Metabolon data did not need to be filtered by coefficient of variation, as there were no technical replicates.  The resulting data consisted of 536 metabolites with 132 patient samples.    
The outcome of this model is the tumor diagnosis (normal tissue vs. tumor tissue). This is a discrete phenotype. In this vignette, the foal is to find gene-metabolite pairs that are differentially associated with one another in normal vs. tumor tissue.

# References

1.  Su, G., Burant, C.F., Beecher, C.W., Athey, B.D. and Meng, F. (2011) Integrated metabolome and transcriptome analysis of the NCI60 dataset. BMC bioinformatics, 12, S36.

2.  Reinhold, W. C., Sunshine, M., Liu, H., Varma, S., Kohn, K. W., Morris, J., Doroshow, J., &#38; Pommier, Y. (2012). CellMiner: a web-based suite of genomic and pharmacologic tools to explore transcript and drug patterns in the NCI-60 cell line set. Cancer Research, 72(14), 3499. https://doi.org/10.1158/0008-5472.CAN-12-1370.

3. Terunuma A, Putluri N, Mishra1 P, Math√© EA, Dorsey TH, Yi M, Wallace TA, Issaq HJ, Zhou M, Killian JK, Stevenson HS, Karoly ED, Chan K, Samanta S, Prieto D, Hsu TY.T., Kurley SJ, Putluri V, Sonavane R, Edelman DC, Wulff J, Starks AM, Yang Y, Kittles RA, Yfantis HG, Lee DH, Ioffe OB, Schiff R, Stephens RM, Meltzer PS, Veenstra TD, Westbrook TF, Sreekumar A, and Stefan Ambs S. MYC-driven 2-hydroxyglutarate associates with poor prognosis in breast cancer. J Clin Invest. 2014 Jan 2;124(1):398-412.